algorithms/mopt/mopt_hierarflow_routines.cpp

```diff
@@ -0,0 +1,213 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2022 Ricerca Security
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+#include "fuzzuf/algorithms/mopt/mopt_hierarflow_routines.hpp"
+
+#include "fuzzuf/algorithms/mopt/mopt_optimizer.hpp"
+#include "fuzzuf/algorithms/mopt/mopt_option.hpp"
+#include "fuzzuf/algorithms/mopt/mopt_option_get_splice_cycles.hpp"
+#include "fuzzuf/utils/common.hpp"
+
+namespace fuzzuf::algorithm::mopt::routine {
+
+namespace other {
+
+using mopt::option::MOptTag;
+
+MOptUpdate::MOptUpdate(MOptState &state) : state(state) {}
+
+MOptMidCalleeRef MOptUpdate::operator()(
+    [[maybe_unused]] std::shared_ptr<MOptTestcase> testcase) {
+  // retrive optimizer values
+  auto last_splice_cycle = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::LastSpliceCycle, true);
+  auto havoc_operator_finds = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::HavocOperatorFinds, true);
+
+  auto &mopt = state.mopt;
+
+  // update havoc_operator_finds
+  for (size_t i = 0; i < havoc_operator_finds.size(); i++) {
+    mopt->havoc_operator_finds[state.core_mode ? 1 : 0][i] +=
+        havoc_operator_finds[i];
+  }
+
+  if (last_splice_cycle >= afl::option::GetSpliceCycles(state)) {
+    state.UpdateSpliceCycles();
+  }
+
+  if (state.pacemaker_mode) {
+    // TODO
+  }
+
+  auto new_testcases = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::NewTestcases, true);
+  auto selected_case_histogram = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::SelectedCaseHistogram, true);
+
+  // pilot mode (update local best)
+  if (!state.core_mode) {
+    if (unlikely(new_testcases > mopt::option::GetPeriodPilot<MOptTag>())) {
+      for (size_t i = 0; i < selected_case_histogram.size(); i++) {
+        double score = 0.0;
+        if (selected_case_histogram[i] > 0) {
+          score = (mopt->havoc_operator_finds[0][i] +
+                   mopt->havoc_operator_finds[1][i]) /
```

Comment:
Here the same as
https://github.com/fuzzuf/fuzzuf/pull/84/files#r934244591

Moreover, it is not obvious for those who are not familiar with MOpt that, if the current mode is not "Core Mode", then it should be "Pilot Mode".
Although the top comment appropriately explain it, I think for example `state.mode == PILOT_MODE` would express it more directly

Reply:
fiexed in 896319460a5f1cd0bb1c0934868b610193412ccf